presto-main/src/main/java/com/facebook/presto/execution/MemoryTrackingRemoteTaskFactory.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution;
+
+import com.facebook.presto.OutputBuffers;
+import com.facebook.presto.Session;
+import com.facebook.presto.execution.StateMachine.StateChangeListener;
+import com.facebook.presto.metadata.Split;
+import com.facebook.presto.spi.Node;
+import com.facebook.presto.sql.planner.PlanFragment;
+import com.facebook.presto.sql.planner.plan.PlanNodeId;
+import com.google.common.collect.Multimap;
+
+public class MemoryTrackingRemoteTaskFactory
+        implements RemoteTaskFactory
+{
+    private final RemoteTaskFactory remoteTaskFactory;
+    private final QueryStateMachine stateMachine;
+
+    public MemoryTrackingRemoteTaskFactory(RemoteTaskFactory remoteTaskFactory, QueryStateMachine stateMachine)
+    {
+        this.remoteTaskFactory = remoteTaskFactory;
+        this.stateMachine = stateMachine;
+    }
+
+    @Override
+    public RemoteTask createRemoteTask(Session session,
+            TaskId taskId,
+            Node node,
+            PlanFragment fragment,
+            Multimap<PlanNodeId, Split> initialSplits,
+            OutputBuffers outputBuffers)
+    {
+        RemoteTask task = remoteTaskFactory.createRemoteTask(session,
+                taskId,
+                node,
+                fragment,
+                initialSplits,
+                outputBuffers);
+
+        task.addStateChangeListener(new UpdatePeakMemory(stateMachine));
+        return task;
+    }
+
+    private static final class UpdatePeakMemory
```

Comment:
this class needs to be thread safe, as the callback may be called from different threads